poetapp/node
# Replace Insight with Bitcoin Core in BlockchainReader

[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

Fixes https://github.com/poetapp/node/issues/106
Part of https://github.com/poetapp/node/issues/24

RDD: [Replacing Insight API with Bitcoin Core](https://github.com/poetapp/rdd/blob/master/replace-insight-with-bitcoind.md).

## Description

This PR completely drops the Insight API dependency for the BlockchainReader module, replacing it with Bitcoin Core.

The Helpers/Bitcoin file now only contains interfaces and enums used with Bitcoin Core's RPC, and the declaration of the two prefixes we support.

Bitcoin-related functions that are specific to the BlockchainReader now live in BlockchainReader/Bitcoin.

## New Configuration Options

Configuration options for Bitcoin Core RPC have been added.

## New Dependencies


I'm using Rui Marinho's [bitcoin-core](https://github.com/ruimarinho/bitcoin-core), a simple wrapper over the Bitcoin Core RPC, with custom TypeScript types for it.

186 repositories currently [depend on ruimarinho/bitcoin-core](https://github.com/ruimarinho/bitcoin-core/network/dependents).

## Tests

### Old Tests

Two test files needed to be updated to reflect the changes, but I decided to delete them instead. They weren't really testing anything — they were originally added to circumvent the coverage report failing to correctly calculate coverage for TypeScript files.

### New Unit Tests

Almost all logic is written in pure functions in BlockchainReader/Bitcoin and unit tested.

I've added a real bitcoin block as json for this purpose.

### Functional Tests

This change can only be correctly tested with [Full Node Syncing](https://github.com/poetapp/node/issues/22), which is blocked by this PR.

### Docker Image of Bitcoin Core

I've added a bitcoin-core entry to docker-compose.yml.

I'm running this with `sudo docker-compose up -d poet-bitcoind`.

It's configured to match the node's default configuration.

### Manual Testing

- Set `"forceBlockHeight": 1411304`
- `db.dropDatabase()`
- `npm start`
- `GET http://localhost:18080/works` should return an empty array
- Let it run for a while
- `GET http://localhost:18080/works` should now return 6 works, each with a full timestamp.
- Compare results of `GET http://localhost:18080/works` to [block-1411304.json.tar.gz](https://github.com/poetapp/node/files/2335054/block-1411304.json.tar.gz)

More and different manual testing is welcome.

## Future Changes

We're currently passing `username` and `password` to the Bitcoin Core RPC. This method is unsafe and deprecated, soon to be removed. We should use the `auth` option instead, which supports passing the hash of the password rather than plaintext.

We'll address this change in a separate PR.
